File: gitahidave/linked_lists.py
# Define the structure of a singly linked list node
class ListNode:
    def __init__(self, value=0, next=None):
        self.value = value
        self.next = next

# Function to reverse the linked list
def reverse_linked_list(head):
    prev = None  # Initialize previous node as None
    current = head  # Start with the head node

    while current:  # Traverse through the linked list
        next_node = current.next  # Save the next node
        current.next = prev  # Reverse the current node's pointer
        prev = current  # Move prev to the current node
        current = next_node  # Move to the next node

    return prev  # prev will be the new head of the reversed list

# Helper function to print the linked list
def print_linked_list(head):
    while head:
        print(head.value, end=" -> ")
        head = head.next
    print("None")

# Example usage:
# Creating a linked list: 1 -> 2 -> 3 -> 4 -> None
head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)

print("Original list:")
print_linked_list(head)

# Reversing the linked list
reversed_head = reverse_linked_list(head)

print("Reversed list:")
print_linked_list(reversed_head)

File: 1anmugambi/linked_lists.js
class ListNode {
    constructor(value = 0, next = null) {
        this.value = value;
        this.next = next;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    // Add node to the end of list
    append(value) {
        if (!this.head) {
            this.head = new ListNode(value);
            return;
        }
        let current = this.head;
        while (current.next) {
            current = current.next;
        }
        current.next = new ListNode(value);
    }

    // Reverse linked list
    reverse() {
        let prev = null;
        let curr = this.head;
        while (curr) {
            let next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        this.head = prev;
    }


    print() {
        let current = this.head;
        let output = [];
        while (current) {
            output.push(current.value);
            current = current.next;
        }
        console.log(output.join(' -> '));
    }
}

// Example
const list = new LinkedList();
list.append(1);
list.append(2);
list.append(3);
list.append(4);
list.append(5);

console.log("Input:");
list.print();

list.reverse();

console.log("Output:");
list.print();